official/vision/beta/projects/yolo/losses/yolo_loss.py

```diff
@@ -0,0 +1,654 @@
+
+    # Stop gradient so while loop is not tracked.
+    obj_mask = tf.stop_gradient(obj_mask)
+    true_conf = tf.stop_gradient(true_conf)
+    return true_conf, obj_mask
+
+  def __call__(self, true_counts, inds, y_true, boxes, classes, y_pred):
+    """Call function to compute the loss and return the total loss as 
+    well as the loss for each detection mask on a given FPN level. 
+    
+    Args: 
+      true_counts: `Tensor` of shape [batchsize, height, width, num_anchors] 
+        represeneting how many boxes are in a given pixel [j, i] in the 
+        output map.
+      inds: `Tensor` of shape [batchsize, None, 3] indicating the location 
+        [j, i] that a given box is associatied with in the FPN prediction 
+        map. 
+      y_true: `Tensor` of shape [batchsize, None, 8] indicating the actual box 
+        associated with each index in the inds tensor list.
+      boxes: `Tensor` of shape [batchsize, None, 4] indicating the original 
+        ground truth boxes for each image as they came from the decoder used
+        for bounding box search. 
+      classes: `Tensor` of shape [batchsize, None, 1] indicating the original 
+        ground truth classes for each image as they came from the decoder used
+        for bounding box search.
+      y_pred: `Tensor` of shape [batchsize, height, width, output_depth] 
+        holding the models output at a specific FPN level.
+
+    Return:
+      loss: `float` for the actual loss.
+      box_loss: `float` loss on the boxes used for metrics.
+      conf_loss: `float` loss on the confidence used for metrics.
+      class_loss: `float` loss on the classes used for metrics.
+      avg_iou: `float` metric for the average iou between predictions 
+        and ground truth.
+      avg_obj: `float` metric for the average confidence of the model 
+        for predictions.
+    """
+    (loss, box_loss, conf_loss, class_loss, mean_loss, iou, pred_conf, ind_mask,
+     grid_mask) = self.call(true_counts, inds, y_true, boxes, classes, y_pred)
+
+    # Temporary metrics
+    box_loss = tf.stop_gradient(0.05 * box_loss / self._iou_normalizer)
+
+    # Metric compute using done here to save time and resources.
+    sigmoid_conf = tf.stop_gradient(tf.sigmoid(pred_conf))
+    iou = tf.stop_gradient(iou)
+    avg_iou = loss_utils.avgiou(
+        loss_utils.apply_mask(tf.squeeze(ind_mask, axis=-1), iou))
+    avg_obj = loss_utils.avgiou(tf.squeeze(sigmoid_conf, axis=-1) * grid_mask)
+    return (loss, box_loss, conf_loss, class_loss, mean_loss,
+            tf.stop_gradient(avg_iou), tf.stop_gradient(avg_obj))
+
+  @abc.abstractmethod
+  def _build_per_path_attributes(self):
+    """Additional initialization required specifically for each unique YOLO 
+    loss version"""
+    ...
+
+  @abc.abstractmethod
+  def call():
+    """The actual logic to apply to the raw model for optimization."""
+    ...
+
+  def post_path_aggregation(self, loss, ground_truths, predictions):
```

Comment:
I can't see the implementation, does this function pass through the loss as is?

Reply:
In most cases it will propagate as is, but if you take a look at ScaledLoss line 510, after the loss has been aggregated, it needs to be scaled by a ratio of the number of FPN paths. So this function is needed for that scaling.  In the DarkNet Loss we do not need to scale by the number of FPN paths, so it follows the default behavior simply by passing through the function.  Taking a look at the YoloX paper down the line, the function will again be needed for loss scaling after aggregation.

Reply:
I see, can you add the default behavior to the docstr?

Reply:
yep added